File: myLeetCode/Binary_Tree/Traversal mode of Binary tree/589. N-ary Tree Preorder Traversal.cpp
// 589. N-ary Tree Preorder Traversal
// https://leetcode.com/problems/n-ary-tree-preorder-traversal/

#include <iostream>
#include <vector>
#include <stack>

// N叉树节点定义
class Node {
public:
    int val;
    std::vector<Node*> children;

    Node() {}
    
    Node(int _val) {
        val = _val;
    }

    Node(int _val, std::vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

// 前序遍历 N 叉树（迭代版本）
std::vector<int> preorder(Node* root) {
    std::vector<int> result;
    if (root == nullptr) {
        return result;
    }
    
    std::stack<Node*> stk;
    stk.push(root);
    
    while (!stk.empty()) {
        Node* curr = stk.top();
        stk.pop();
        
        result.push_back(curr->val);
        
        for (int i = curr->children.size() - 1; i >= 0; i--) {
            stk.push(curr->children[i]);
        }
    }
    
    return result;
}

int main() {
    // 构建测试 N 叉树
    Node* root = new Node(1);
    Node* node2 = new Node(2);
    Node* node3 = new Node(3);
    Node* node4 = new Node(4);
    Node* node5 = new Node(5);
    Node* node6 = new Node(6);
    root->children.push_back(node3);
    root->children.push_back(node2);
    root->children.push_back(node4);
    node3->children.push_back(node5);
    node3->children.push_back(node6);

    // 前序遍历 N 叉树
    std::vector<int> result = preorder(root);

    // 打印遍历结果
    std::cout << "Preorder traversal: ";
    for (int val : result) {
        std::cout << val << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: C#/myLeetCode/Binary_Tree/N-ary Tree/589. N-ary Tree Preorder Traversal.cs
// 589. N-ary Tree Preorder Traversal
// https://leetcode.com/problems/n-ary-tree-preorder-traversal/

using System;
using System.Collections.Generic;

public class Node {
    public int val;
    public IList<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
        children = new List<Node>();
    }

    public Node(int _val, IList<Node> _children) {
        val = _val;
        children = _children;
    }
}

public class Solution {
    public IList<int> Preorder(Node root) {
        List<int> result = new List<int>();
        PreorderTraversal(root, result);
        return result;
    }

    private void PreorderTraversal(Node node, List<int> result) {
        if (node == null) {
            return;
        }

        result.Add(node.val);

        foreach (var child in node.children) {
            PreorderTraversal(child, result);
        }
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();

        Node tree = new Node(1, new List<Node> {
            new Node(3, new List<Node> {
                new Node(5),
                new Node(6)
            }),
            new Node(2),
            new Node(4)
        });

        IList<int> preorderTraversal = solution.Preorder(tree);

        Console.WriteLine("Preorder Traversal:");
        foreach (int val in preorderTraversal) {
            Console.Write(val + " ");
        }
    }
}
